database/postgresql/src/main/java/com/example/javaopencommerce/catalog/CatalogRepositoryImpl.java

```diff
@@ -0,0 +1,54 @@
+package com.example.javaopencommerce.catalog;
+
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+class CatalogRepositoryImpl implements CatalogRepository {
+
+  private final PsqlCategoryRepository categoryRepository;
+  private final ObjectMapper objectMapper;
+
+  CatalogRepositoryImpl(PsqlCategoryRepository categoryRepository, ObjectMapper objectMapper) {
+    this.categoryRepository = categoryRepository;
+    this.objectMapper = objectMapper;
+  }
+
+  @Override
+  public Category getCatalog() {
+    String catalogJson = categoryRepository.getCatalog();
+    return recoverCategory(catalogJson);
+  }
+
+  @Override
+  public Category saveCatalog(Category catalog) {
+    Category.CategorySnapshot catalogSnapshot = catalog.toSnapshot();
+    CategoryEntity catalogEntity = CategoryEntity.fromSnapshot(catalogSnapshot);
+
+    String catalogJson;
+    try {
+      catalogJson = objectMapper.writeValueAsString(catalogEntity);
+    } catch (JsonProcessingException e) {
+      throw new IllegalStateException("Problems with category tree save/update", e);
+    }
+
+    catalogJson = categoryRepository.saveCatalog(catalogJson);
+    return recoverCategory(catalogJson);
+  }
+
+  private Category recoverCategory(String catalogJson) {
```

Comment:
Please rename it to 'parseCategory' or something. It's not "recovery", as this word implicates that something bad happen, and we have to handle it in order to fix or at least try to fix it.